cEP-0021.md

```diff
@@ -0,0 +1,379 @@
+| Status   | Proposed                                          |
+| Type     | Process                                           |
+
+## Abstract
+
+This cEP describes the detailed implementation of profiling interface as a part
+of the
+[GSoC'18 project](https://summerofcode.withgoogle.com/projects/#6109762077327360).
+
+## Introduction
+
+Profiling a Python program is doing a dynamic analysis that measures the
+execution time of the program and everything that compose it. That means
+measuring the time spent in each of its functions. This will give the data about
+where our program is spending time, and what area might be worth optimizing,
+thus we can provide an *Profiling interface* from within coala and make it
+available to bear writers so that they can create a Better Performant bears.
+
+This proposal is to introduce the Profiling interface so that bear writer will
+have the ability to jump directly into Bear code and move through it using a
+debugger interface and Profile the Bear code to create a Better Performant
+bears.
+
+## Proposed Change
+
+Bear writer can invoke Profiler through *CLI*:
+
+- By passing custom settings through *CLI*
+  `coala -b Bear Mybear --files file --profile-bears -"True,...."`,
+- By using coala default settings provided by coala
+  `coala -b Bear Mybear --files file --profile-bears`,the default are to strip
+  the directory (`strip_dirs()`) and sort the stats of profiler result by
+  cummulative time (`cumtime`) and display the profiler result on *CLI*
+
+Bear writer can invoke Profiler by using `coafile` i.e. `profile_bears = XXX`.
+
+- The profiling will start by profiling the `run()` method of the bears.
+  This will load the settings passed to it, and then run the Profiler with the
+  passed settings on `run()` method of Bear.
+
+- The advantage of profiling just the `run()` method of the bear is that,
+  bear writer doesn't have to spend time on loading the files,
+  collecting the settings, etc. as this is all done by coala itself, they just
+  have to Profile the `run()` method of Bear to create a Better PerformantBear.
+
+After invoking the profiler on the run method of a bear it will first save
+all the profile output on a local temp file which further user by
+`pstats module` to profile results and formats them in a way Bear
+writer wants to format it.
+
+> Currently default format is to strip the directory (`strip_dirs()`) and sort
+> the stats of profiler result by cummulative time (`cumtime`) and display the
+> profiler result on *CLI*.
+>
+> ```python
+> pstats.strip_dirs().sort_stats('cumtime')
+> ```
+
+After formatting the profile output Bear writer will have an option to
+either display the formatted ouput on terminal or dump the formatted result
+in a specified file.
+
+```
+coala -b Bear Mybear --files <file> --profile-bears "<filename or file-path>"
+```
+
+Bear writer will also have a option to direclty dump the profile output
```

Comment:
>Bear writer will also have a option to direclty dump the profile outpu

unnecessary detail. all the bear writer is concerned with is getting the sunburst and icicle diagrams in a visual way